app/server/api2/geolocation_api.py

```diff
@@ -0,0 +1,237 @@
+
+
+class LocationAPI(MethodView):
+    """Handles requests that operate on external location references
+
+    Documentation is brief since it implmements standard flask blueprint interface
+    """
+
+    # TODO: This method is too long and should be divided up for clarity
+    def get(self, path_string):
+        """Returns the location entry matching the path string
+        """
+        # split given path into array elements
+        # validate that each part is a valid location name token
+        path_parts = path.reverse_split(path_string)
+        c = 0
+        for path_part in path_parts:
+            if path_part == '':
+                break
+            if valid_location_name(path_part):
+                c += 1
+        if c == 0:
+            if not valid_location_name(path_part):
+                response_object = {
+                    'message': 'Invalid location path: {}'.format(common_name)
+                }
+                return make_response(jsonify(response_object)), 400
+  
+
+        # if one element is found, and if a path with more than one element is given
+        # make sure all the elements in the hierarchy match
+        common_name = path_parts[0]
+        locations = Location.query.filter(Location.common_name==common_name).all()
+        valid_locations = []
+        valid = True
+        while len(locations) > 0:
+            location = locations.pop(0)
+            location_step = location
+            for part in path_parts:
+                if location_step.parent == None:
+                    if len(valid_locations) > 0:
+                        valid = False
+                    break
+                # TODO: fuzzy match
+                if part != location_step.common_name:
+                    valid = False
+                    break
+                location_step = location_step.parent
+            if valid:
+                valid_locations.append(location)
+
+        if len(valid_locations) == 0:
+            response_object = {
+                    'message': 'Location path {} not found'.format(path_string),
+                }
+            return make_response(jsonify(response_object)), 404
+
+        response_object = {
+            'search_string': common_name,
+            'locations': []
+                }
+        for location in valid_locations:
+            response_object['locations'].append({
+                'id': location.id,
+                'common_name': location.common_name,
+                'path': str(location),
+                'latitude': location.latitude,
+                'longitude': location.longitude,
+                }
+                )
+
+        return make_response(jsonify(response_object)), 200
+
+
+
+    # TODO: This method is too long and should be divided up for clarity
+    def post(self):
+        """Commits the given location entry to the database.
+
+        Will fail if the location name and path already exist, or if external reference already exists.
```

Comment:
If it will fail for this reason shouldn't there be a check for whether the location entry exits and a subsequent response matching that error; something along the lines of:
```
existing_location = Location.query.filter_by(latitute=lattitude, longitute=longitude).first()
if existing_location:
    response = {"error": "Location already exists."}
```

Reply:
@PhilipWafula you are right, error handling is bad here. I will fix it.